s3driver_test.go

```diff
@@ -0,0 +1,241 @@
+
+	mockS3API.EXPECT().DeleteObject(&s3.DeleteObjectInput{
+		Bucket: aws.String("bucket"),
+		Key:    aws.String("home/dir/"),
+	}).Return(nil, nil)
+
+	driver := &S3Driver{
+		s3:       mockS3API,
+		bucket:   "bucket",
+		homePath: "home",
+	}
+	err := driver.DeleteDir("../../dir/")
+
+	assert.NoError(t, err)
+}
+
+func TestDeleteFile(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockS3API := NewMockS3API(mockCtrl)
+
+	mockS3API.EXPECT().DeleteObject(&s3.DeleteObjectInput{
+		Bucket: aws.String("bucket"),
+		Key:    aws.String("home/dir/file"),
+	}).Return(nil, nil)
+
+	driver := &S3Driver{
+		s3:       mockS3API,
+		bucket:   "bucket",
+		homePath: "home",
+	}
+	err := driver.DeleteDir("../../dir/file")
+
+	assert.NoError(t, err)
+}
+
+func TestRename(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockS3API := NewMockS3API(mockCtrl)
+
+	mockS3API.EXPECT().CopyObject(&s3.CopyObjectInput{
+		Bucket:     aws.String("bucket"),
+		CopySource: aws.String("bucket/home/dir/file"),
+		Key:        aws.String("home/dir/new_file"),
+	}).Return(nil, nil)
+
+	mockS3API.EXPECT().DeleteObject(&s3.DeleteObjectInput{
+		Bucket: aws.String("bucket"),
+		Key:    aws.String("home/dir/file"),
+	}).Return(nil, nil)
+
+	driver := &S3Driver{
+		s3:       mockS3API,
+		bucket:   "bucket",
+		homePath: "home",
+	}
+	err := driver.Rename("dir/file", "dir/new_file")
+
+	assert.NoError(t, err)
+}
+
+func TestMakeDir(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockS3API := NewMockS3API(mockCtrl)
+
+	mockS3API.EXPECT().PutObject(&s3.PutObjectInput{
+		Bucket: aws.String("bucket"),
+		Key:    aws.String("home/new_dir/"),
+		Body:   bytes.NewReader([]byte{}),
+	}).Return(nil, nil)
+
+	driver := &S3Driver{
+		s3:       mockS3API,
+		bucket:   "bucket",
+		homePath: "home",
+	}
+	err := driver.MakeDir("/new_dir")
+
+	assert.NoError(t, err)
+}
+
+func TestGetFile(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockS3API := NewMockS3API(mockCtrl)
+
+	mockS3API.EXPECT().GetObject(&s3.GetObjectInput{
+		Bucket: aws.String("bucket"),
+		Key:    aws.String("home/dir/file"),
+	}).Return(&s3.GetObjectOutput{
+		Body: nil,
+	}, nil)
+
+	driver := &S3Driver{
+		s3:       mockS3API,
+		bucket:   "bucket",
+		homePath: "home",
+	}
+	_, err := driver.GetFile("../../dir/file")
+
+	assert.NoError(t, err)
+}
+
+func TestPutFile(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockS3API := NewMockS3API(mockCtrl)
+
+	mockS3API.EXPECT().PutObject(&s3.PutObjectInput{
+		Bucket: aws.String("bucket"),
+		Key:    aws.String("home/dir/file"),
+		Body:   bytes.NewReader([]byte{1, 2, 3}),
+	}).Return(nil, nil)
+
+	driver := &S3Driver{
+		s3:       mockS3API,
+		bucket:   "bucket",
+		homePath: "home",
+	}
+	err := driver.PutFile("../../dir/file", bytes.NewReader([]byte{1, 2, 3}))
+
+	assert.NoError(t, err)
+}
```

Comment:
I think this is one module where testing against actual S3 would be beneficial (as opposed to mocks)